### PR TITLE
Support Continuation Object Scanning

### PR DESCRIPTION
**Pull request by LinHu2016**
- add RootScannerEntity_ContinuationObjects
	- add _continuationObjectCount in AllocationStats
	- add continuations element in verbosegc schema
	- new MM_ScavengeScanReason
	- new HeapWalkerDelegate to handle specific scan for Continuation
	- update ScavengeDelegate interface to handle specific scan for
	Continuation.

Signed-off-by: Lin Hu <linhu@ca.ibm.com>

### CODE REVIEW

**Comment by LinHu2016** [comment]
@amicic @dmitripivkine please review the change, Thanks

File: gc/base/standard/HeapWalker.cpp [hunk 1]
```diff
@@ -70,14 +71,23 @@ heapWalkerObjectFieldSlotDo(OMR_VM *omrVM, omrobjectptr_t object, GC_SlotObject
...
+heapWalkerObjectSlotDo(OMR_VM *omrVM, OMR_VMThread *omrVMThread, omrobjectptr_t object, MM_HeapWalkerSlotFunc oSlotIterator, void *localUserData, MM_HeapWalkerDelegate *delegate)
 {
 	GC_ObjectIterator objectIterator(omrVM, object);
 	GC_SlotObject *slotObject;
 
 	while ((slotObject = objectIterator.nextSlot()) != NULL) {
 		heapWalkerObjectFieldSlotDo(omrVM, object, slotObject, oSlotIterator, localUserData);
 	}
+
+	delegate->objectSlotDo(omrVMThread, object, oSlotIterator, localUserData);
```
Comment (amicic) [inline gc/base/standard/HeapWalker.cpp:83]: objectSlotDo -> objectSlotsDo

File: gc/base/standard/Scavenger.cpp [hunk 2]
```diff
@@ -2776,8 +2776,12 @@ MM_Scavenger::shouldRememberObject(MM_EnvironmentStandard *env, omrobjectptr_t o
 	 * So we must not split, if it's indexable
 	 */
 	uintptr_t scannerFlags = GC_ObjectScanner::scanRoots | GC_ObjectScanner::indexableObjectNoSplit;
+	bool shouldRemember = false;
 
-	GC_ObjectScanner *objectScanner = getObjectScanner(env, objectPtr, &objectScannerState, scannerFlags);
+	GC_ObjectScanner *objectScanner = getObjectScanner(env, objectPtr, &objectScannerState, scannerFlags, SCAN_REASON_SHOULDREMEMBER, &shouldRemember);
+	if (shouldRemember) {
```
Comment (amicic) [inline gc/base/standard/Scavenger.cpp:2782]: now that you introduced flag shouldRemember, let's try to make the method does only one 'return shouldRemember' at the end, and whatever other changes needs to be done to eliminate other returns

File: include_core/omrgcconsts.h [hunk 3]
```diff
@@ -121,6 +121,14 @@ typedef enum MM_MarkingSchemeScanReason {
 	SCAN_REASON_OVERFLOWED_OBJECT = 4, /**< Indicates the object being scanned was in an overflowed region */
 } MM_MarkingSchemeScanReason;
 
+typedef enum MM_ScavengeScanReason {
+	SCAN_REASON_SCAVENGE = 10,
+	SCAN_REASON_INCREMENTALSCAVENGE = 11,
```
Comment (amicic) [inline include_core/omrgcconsts.h:126]: SCAN_REASON_INCREMENTALSCAVENGE is probably no different than SCAN_REASON_SCAVENGE and could be removed

File: gc/base/standard/Scavenger.cpp (hunk 2 above)
Comment (LinHu2016) [reply gc/base/standard/Scavenger.cpp:2797]: Do you mean that we change the method like `shouldRemember = scanObject(....);` and delegate will handle the whole scanning via ObjectScanner internally? Thanks

File: gc/base/MarkingScheme.cpp [hunk 4]
```diff
@@ -429,6 +429,25 @@ MM_MarkingScheme::fixupForwardedSlotOutline(GC_SlotObject *slotObject) {
...
+	bool const compressed = _extensions->compressObjectReferences();
+	if (_extensions->getGlobalCollector()->isStwCollectionInProgress()) {
+		MM_ForwardedHeader forwardHeader(*slotPtr, compressed);
+		omrobjectptr_t forwardPtr = forwardHeader.getNonStrictForwardedObject();
+
+		if (NULL != forwardPtr) {
+			if (forwardHeader.isSelfForwardedPointer()) {
+				forwardHeader.restoreSelfForwardedPointer();
+			} else {
+				*slotPtr =forwardPtr;
```
Comment (amicic) [inline gc/base/MarkingScheme.cpp:444]: space missing: '=  forwardPtr'

File: gc/base/MarkingScheme.cpp [hunk 5]
```diff
@@ -429,6 +429,25 @@ MM_MarkingScheme::fixupForwardedSlotOutline(GC_SlotObject *slotObject) {
...
+			}
+		}
+	}
+#endif /* OMR_GC_CONCURRENT_SCAVENGER */
+}
+
```
Comment (amicic) [inline gc/base/MarkingScheme.cpp:450]: can the existing method be rewritten to share to code with this new method to be like this:

```
bool
fixupForwardedSlotOutline(GC_SlotObject *slotObject)
{
	omrobjectptr_t slot = slotObject->readReferenceFromSlot();
	fixupForwardedSlotOutline(env, slot);
	slotObject->writeReferenceToSlot(slot);
}

Comment (amicic) [reply gc/base/MarkingScheme.cpp:450]: ... and possibly move it to hpp and inline it

File: gc/base/standard/Scavenger.cpp [hunk 6]
```diff
@@ -3938,12 +3958,31 @@ MM_Scavenger::fixupSlot(GC_SlotObject *slotObject)
...
+			*slotPtr = forwardHeader.getForwardedObject();
+			Assert_MM_false(isObjectInEvacuateMemory(*slotPtr));
+			return true;
+		} else {
+			Assert_MM_false(_extensions->objectModel.isDeadObject(objectPtr));
+		}
+	}
+	return false;
+}
+
```
Comment (amicic) [inline gc/base/standard/Scavenger.cpp:3978]: can we share the code by rewriting the original method to use this one, and move it to hpp and inline it?

```
bool
MM_Scavenger::fixupSlot(GC_SlotObject *slotObject)
{
	omrobjectptr_t objectPtr = slotObject->readReferenceFromSlot();
	if (fixupSlot(&objectPtr)) {
		slotObject->writeReferenceToSlot(objectPtr);
		return true;
	}
	return false;
}
```

File: gc/base/standard/Scavenger.cpp (hunk 2 above)
Comment (amicic) [reply gc/base/standard/Scavenger.cpp:2797]: nah, slot iteration loop should stay here. I was just hoping that number of returns could be reduced, but probably can't be much if any better than this. leave it as is

File: example/glue/ScavengerDelegate.hpp [hunk 7]
```diff
@@ -131,10 +131,12 @@ class MM_ScavengerDelegate : public MM_BaseVirtual {
 	 * @param[in] objectPtr The object to be scanned
 	 * @param[in] allocSpace Space for in-place instantiation of scanner
 	 * @param[in] flags See GC_ObjectScanner::InstanceFlags. One of scanRoots or scanHeap will be set , but not both.
+	 * @param[in] reason See MM_ScavengeScanReason
+	 * @param[out]shouldRemember
 	 * @return Pointer to object scanner, or NULL if object not to be scanned (eg, leaf object).
 	 * @see GC_ObjectScanner
 	 */
-	GC_ObjectScanner *getObjectScanner(MM_EnvironmentStandard *env, omrobjectptr_t objectPtr, void *allocSpace, uintptr_t flags);
+	GC_ObjectScanner *getObjectScanner(MM_EnvironmentStandard *env, omrobjectptr_t objectPtr, void *allocSpace, uintptr_t flags, MM_ScavengeScanReason reason, bool *shouldRemember);
```
Comment (amicic) [inline example/glue/ScavengerDelegate.hpp:139]: don't you need to leave the old API temporarily and remove it later after delegates in downstream projects are updated ?

File: gc/base/standard/HeapWalker.cpp [hunk 8]
```diff
@@ -70,34 +71,44 @@ heapWalkerObjectFieldSlotDo(OMR_VM *omrVM, omrobjectptr_t object, GC_SlotObject
...
 	}
+
+	delegate->objectSlotsDo(omrVMThread, object, oSlotIterator, localUserData);
+}
+
+void
+MM_HeapWalker::heapWalkerSlotCallback(MM_EnvironmentBase *env, omrobjectptr_t *objectSlotPtr, MM_HeapWalkerSlotFunc function, void * userData)
+{
+	OMR_VM *omrVM = env->getOmrVMThread()->_vm;
+	(*function)(omrVM, objectSlotPtr, userData,0);
```
Comment (amicic) [inline gc/base/standard/HeapWalker.cpp:90]: space missing: 'userData, 0'

File: gc/base/standard/HeapWalker.cpp [hunk 9]
```diff
@@ -70,34 +71,44 @@ heapWalkerObjectFieldSlotDo(OMR_VM *omrVM, omrobjectptr_t object, GC_SlotObject
  * walk through slots of mixed object and apply the user function.
  */
 static void
-heapWalkerObjectSlotDo(OMR_VM *omrVM, omrobjectptr_t object, MM_HeapWalkerSlotFunc oSlotIterator, void *localUserData)
+heapWalkerObjectSlotsDo(OMR_VM *omrVM, OMR_VMThread *omrVMThread, omrobjectptr_t object, MM_HeapWalkerSlotFunc oSlotIterator, void *localUserData, MM_HeapWalkerDelegate *delegate)
```
Comment (amicic) [inline gc/base/standard/HeapWalker.cpp:74]: not that you pass OMR_VMThread, don't pass OMR_VM, but get it from OMR_VMThread

File: example/glue/ScavengerDelegate.hpp (hunk 7 above)
Comment (LinHu2016) [reply example/glue/ScavengerDelegate.hpp:139]: in that case, we need 2 merges in omr and 2 merges in openJ9(at least one merges)

File: gc/base/MarkingScheme.hpp [hunk 10]
```diff
@@ -296,7 +296,13 @@ class MM_MarkingScheme : public MM_BaseVirtual
 #endif /* OMR_GC_CONCURRENT_SCAVENGER */
 	}
 	
-	void fixupForwardedSlotOutline(GC_SlotObject *slotObject);
+	void fixupForwardedSlotOutline(omrobjectptr_t *slotPtr);
+	MMINLINE void fixupForwardedSlotOutline(GC_SlotObject *slotObject)
+	{
+		omrobjectptr_t slot = slotObject->readReferenceFromSlot();
+		fixupForwardedSlotOutline(env, &slot);
```
Comment (amicic) [inline gc/base/MarkingScheme.hpp:303]: `env` is not need

File: gc/base/MarkingScheme.hpp [hunk 11]
```diff
@@ -296,7 +296,13 @@ class MM_MarkingScheme : public MM_BaseVirtual
 #endif /* OMR_GC_CONCURRENT_SCAVENGER */
 	}
 	
-	void fixupForwardedSlotOutline(GC_SlotObject *slotObject);
+	void fixupForwardedSlotOutline(omrobjectptr_t *slotPtr);
+	MMINLINE void fixupForwardedSlotOutline(GC_SlotObject *slotObject)
```
Comment (amicic) [inline gc/base/MarkingScheme.hpp:300]: this method should be not be called Outline anymore

**Review by amicic — APPROVED** [review]

**Comment by babsingh** [comment]
jenkins build all

**Comment by LinHu2016** [comment]
the error on macOS is fvtest/compilerunittest error, looks like related with JIT, https://dev.azure.com/eclipse-omr/ea4519db-b27e-4d19-a971-f01491f803e3/_apis/build/builds/6267/logs/44

**Comment by LinHu2016** [comment]
I will try to rebase the PR with master branch.

**Comment by babsingh** [comment]
> https://dev.azure.com/eclipse-omr/ea4519db-b27e-4d19-a971-f01491f803e3/_apis/build/builds/6267/logs/44

Known failure: https://github.com/eclipse/omr/issues/6556 (first time seen on macOS)

```
2022-08-19T19:47:24.7540520Z 46: [0;32m[----------] [m16 tests from AbsVPValueTest
2022-08-19T19:47:24.7552840Z 44/46 Test #46: compunittest ........................***Exception: SegFault  0.19 sec
```

> https://ci.eclipse.org/omr/job/PullRequest-linux_ppc-64_le_gcc/3773/console

Known failure: https://github.com/eclipse/omr/issues/6556 

```
16:54:34  30: [0;32m[----------] [m16 tests from AbsVPValueTest
16:54:34  30/30 Test #30: compunittest ......................***Exception: SegFault  0.28 sec
```

> https://ci.eclipse.org/omr/job/PullRequest-zos_390-64/3774/console

Infra issue related to Github. Will restart.

**Comment by babsingh** [comment]
> https://github.com/eclipse/omr/pull/6575/commits

The commit message has unneeded indentation. Can it be removed?

```
	- add RootScannerEntity_ContinuationObjects
	- add _continuationObjectCount in AllocationStats
	- add continuations element in verbosegc schema
	- new MM_ScavengeScanReason
	- new HeapWalkerDelegate to handle specific scan for Continuation
	- update ScavengeDelegate interface to handle specific scan for
	Continuation.
```

**Comment by babsingh** [comment]
> https://hyc-runtimes-jenkins.swg-devops.com/job/Build_JDK11_ppc64_aix_Personal/935/

All OpenJ9 builds, launched by @LinHu2016, are failing. Error from one of the builds:

```
18:10:53  /home/jenkins/workspace/Build_JDK19_ppc64_aix_Personal/omr/gc/base/standard/HeapWalker.hpp:31:10: fatal error: 'HeapWalkerDelegate.hpp' file not found
18:10:53  #include "HeapWalkerDelegate.hpp"
18:10:53           ^~~~~~~~~~~~~~~~~~~~~~~~
18:10:53      1500-036: (I) The NOSTRICT option (default at OPT(3)) has the potential to alter the semantics of a program.  Please refer to documentation on the STRICT/NOSTRICT option for more information.
```

This PR will require some more work before it can be merged.

**Comment by babsingh** [comment]
On leave till 31 Aug 22; won't be able to complete the review. Delegating the review to @dsouzai.

**Comment by LinHu2016** [comment]
Thanks @babsingh, There was a new delagate API, which depends OpenJ9 changes, it need to two stage merge, I have updated this PR for the first stage merge (https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/13946/), @dsouzai Could you please merge this PR? Thanks

**Comment by dsouzai** [comment]
@amicic could you please review the most recent force pushes, namely https://github.com/eclipse/omr/compare/da6513e3d7181b895fae5bb1c1d47628dae78eb4..001ad68b9bc33a1e7f1afba12311fb9f3709431c & https://github.com/eclipse/omr/compare/001ad68b9bc33a1e7f1afba12311fb9f3709431c..3e577cfdfc125e80c9fae487934b06c35e50aba4

I have no knowledge in this area so I need an SME to vet the most recent updates.

**Comment by dsouzai** [comment]
jenkins build all

**Comment by amicic** [comment]
@dsouzai This change is a part of larger set of changes, with 2 way dependency between downstream projects and OMR. Once this is merged and all downstream projects introduce missing entities, this code will be uncommented in a follow-up OMR PR.

While this PR could have gone without these 'comments', I'm ok with them, since we know we'll have to 'activate' that code eventually to get the all functionality, and on the other side its presence helped me get a better overall picture of all the changes.

Also, we could have introduce some of those entities in downstream projects before these OMR, changes, but (due to 2 way dependency) we would face missing OMR functionality and temporary/commented code over there - we just had to start somewhere.

**Comment by dsouzai** [comment]
Test failures are due to https://github.com/eclipse/omr/issues/6556, a previously seen issue (https://github.com/eclipse/omr/pull/6316#issuecomment-1027198038), and infra failures.

**Review by dsouzai — APPROVED** [review]

**Pull request merged**